---
title: "Lamian: a statistical framework for differential pseudotime analysis in multiple single-cell RNA-seq "
author: 
  - name: Wenpin Hou
    affiliation: 
      - &id1 "Deparment of Biostatistics, Bloomberg School of Public Health, Johns Hopkins University"
package: Lamian
output: 
  BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{1. Lamian package}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Overview

Pseudotime analysis based on single-cell RNA-seq (scRNA-seq) data has been widely used to study dynamic gene regulatory programs along continuous biological processes such as cell differentiation, immune responses, and disease development. Existing pseudotime analysis methods primarily address the issue of reconstructing cellular pseudotemporal trajectories and inferring gene expression changes along the reconstructed trajectory in one biological sample. As scRNA-seq studies are increasingly performed on multiple patient samples, comparing gene expression dynamics across samples has been emerging as a new demand for which a systematic analytical solution is lacking. 

We develop a systematic computational and statistical framework, Lamian, for multi-sample pseudotime analysis. Given scRNA-seq data from multiple biological samples with covariates (e.g., age, sex, sample type, disease status, etc.), this framework allows one to (1) construct cellular pseudotemporal trajectory, evaluate the uncertainty of the trajectory branching structure, (2) evaluate changes in branching structure associated with sample covariates, (3) identify changes in cell abundance and gene expression along the pseudotime, and (4) characterize how sample covariates modifies the pseudotemporal dynamics of cell abundance and gene expression. Importantly, when identifying cell abundance or gene expression changes associated with pseudotime and sample covariates, Lamian accounts for variability across biological samples which other existing pseudotime methods do not consider. As a result, Lamian is able to more appropriately control the false discovery rate (FDR) when analyzing multi-sample data, a property not offered by any other methods. We evaluate using synthetic and real scRNA-seq datasets the various tests available in Lamian.

For more details, see our paper describing the `Lamian` package:

- (give paper info here)




```{r}
# load in Lamian
options(warn=-1)
suppressMessages(library(Lamian))
```

# Module 1: tree variability

The module 1 of Lamian is designed for detecting the stability of branches in a pseudotime tree structure. We automatically enumerate all branches from the pseudotime tree structure, and then test for their detection rate through 10,000 bootstraps. After each cell-bootstrapping, we reconstruct the tree structure and re-identify all branches. We apply both Jaccard statistics and overlap coefficient as the statistics for evaluating whether any branch in a bootstrap setting matches with one of the original branches. A branch's detection rate is defined as the percentage of bootstrap settings that a branch finds it match. Module 1 also reports and tests for samples' proportion in each branch. 

### data Preparation

We will need a gene by cell expression matrix, the low-dimension representation of the cells, and the cell annotation (which sample each cell belongs to). Here, we use example data `hca_bm_saver`, `hca_bm_pca`, `hca_bm_cellanno` to demonstrate their data structures. Users can read in their own data of interest in this step.

```{r}
data(hca_bm_saver)
data(hca_bm_pca)
data(hca_bm_cellanno)
```

`hca_bm_saver` is a gene by cell expression matrix. 
```{r}
str(hca_bm_saver)
```

`hca_bm_saver`, `hca_bm_pca` is the low-dimension representation of the cells.
```{r}
str(hca_bm_pca)
```

`hca_bm_cellanno` is the cell annotation where the first column are the cell names, the second column are the sample names, and the row names are the cell names. 
```{r}
str(hca_bm_cellanno)
```

### infer tree structure

```{r}
res = infer_tree_structure(pca = hca_bm_pca, expression = hca_bm_saver, cellanno = hca_bm_cellanno, origin.marker = c('CD34'), xlab='Principal component 1', ylab = 'Principal component 2')
```

As we can see from the above inputs, there are five clusters, and the tree structure inferred based on these clusters consists of three branches: 2 --> 3 --> 1, 2 --> 4, and 2 --> 5. The result object `res` is a list containing information about the tree structure, branches, cell clusters, pseudotime ordering the cells, etc..
```{r}
names(res)
```

### plot the tree structure

```{r fig_plotmclust, fig.height = 5, fig.width = 6, fig.align = "center"}
plotmclust(res, cell_point_size = 0.1, x.lab = 'Principal component 1', y.lab = 'Principal component 2')
```

## evaluate the uncertainty of tree branches
```{r}
result <- evaluate_uncertainty(res, n.permute=3)
names(result)
```

The result is a list of three elements. The first selement is the detection rate of each branch. 
```{r}
result[[1]]
```

The second element is the sample proportion mean information.
```{r}
result[[2]]
```

The thrid element is the sample proportion sd (standard deviation) information. 
```{r}
result[[3]]
```



# Module 2: Trajectory differential tests

## Sample Covariate Test

In the following, we will use an example dataset `expdata` of 100 genes and 1000 cells to demonstration the workflow. In practice, we can input any other interesting dataset. 

```{r}
data(expdata)
```

The inputs should contain: (a) `expr`: a gene by cell expression matrix. Values are library-size-normalized log-transformed gene expression matrix. They can be either imputed or non-imputed. Zero-expression genes should have been filtered out. 
```{r}
str(expdata$expr)
```

(b) `cellanno`: a dataframe where the first column are cell names and second column are sample names. 
```{r}
head(expdata$cellanno)
```

(c) `pseudotime`: a numeric vector of pseudotime, and the names of this vector are the corresponding cell names.
```{r}
str(expdata$pseudotime)
```

(d) `design`: a matrix. Number of rows should be the same as the number of unique samples. Rownames are sample names. First column is the intercept (all 1), second column is the covariate realization valuels for each of the samples. 
```{r}
print(expdata$design)
```

The function ```testpt()``` is designed to perform multiple tests. To perform the Sample Covariate Test, we need to set ```test.type = 'variable'```.

```{r}
Res <- lamian.test(expr = expdata$expr, cellanno = expdata$cellanno, pseudotime = expdata$pseudotime, design = expdata$design, test.type = 'variable')
```

We will know which are diferential dynamic genes (DDG) from the `statistics` data frame in the result object `Res`
```{r}
## get differential dynamic genes statistics 
stat <- Res$statistics
head(stat)
stat <- stat[order(stat[,1], -stat[,3]), ]
diffgene <- rownames(stat[stat[, grep('^fdr.*overall$', colnames(stat))] < 0.05,])
str(diffgene)
```

We will get the population-level estimates for all the DDG by applying function `getPopulationFit()`.
```{r}
## population fit
Res$populationFit <- getPopulationFit(Res, gene = diffgene, type = 'variable')
```

We can apply `getCovariateGroupDiff()` to calculate the group difference regarding this sample covariate, and then cluster the DDG based on the group difference. By setting *k = 4*, we will get two clusters for meanSig DDG, and the other two clusters for DDG of other types.
```{r}
## clustering
Res$covariateGroupDiff <- getCovariateGroupDiff(testobj = Res, gene = diffgene)
Res$cluster <- clusterGene(Res, gene = diffgene, type = 'variable', k=5)
table(Res$cluster)
```

Plot the cluster mean and difference
```{r fig_plotClusterMeanAndDiff, fig.height = 8, fig.width = 3.5, fig.align = "center"}
## plotClusterMeanAndDiff
plotClusterMeanAndDiff(Res)
```
We can also plot the cluster difference seperately by calling `plotClusterDiff(testobj=Res, gene = diffgene)`.

We can further check out whether the DDG in each cluster have any enriched genome functions by applying Gene Ontology (GO) analysis.
```{r, fig_plotGOEnrich, fig.height = 2.6, fig.width = 3, fig.align = "center", eval=FALSE}
## GO analysis
goRes <- GOEnrich(testobj = Res, type = 'variable')  
plotGOEnrich(goRes = goRes, fdr.cutoff = 0.05, fc.cutoff = 2)
```
In practice, if there are significant GO terms in the clusters, the above `plotGOEnrich()` function will generate a heatmap to show the top GO terms, as follows.

![](/Users/wenpinhou/Dropbox/trajectory_variability/package/Lamian/data/hm_GO_term_fdr.png)

### compare original and fitted expression
```{r fig_plotFitHm, fig.height = 8, fig.width = 12, fig.align = "center"}
plotFitHm(Res, type = 'variable', cellWidthTotal = 200, cellHeightTotal = 350)
```

We can also plot original expression values, model fitted expression values, and model-fitted group difference in three seperate heatmaps, as follows.
```{r fig_sct_plotDiffFitHm3, fig.height = 9, fig.width = 16, fig.align = "center"}
plotDiffFitHm3(Res, cellWidthTotal = 180, cellHeightTotal = 350)
```
## Constant Time Test
The inputs for Constant Time Test are the same as those for Sample Covariate Test, except that the `design` matrix can have only one intercept column. If there are more than one columns in `design`, only the first column will be considered. 
```{r}
Res <- lamian.test(expr = expdata$expr, cellanno = expdata$cellanno, pseudotime = expdata$pseudotime, design = expdata$design, test.type = 'time', ncores = 1)
names(Res)
```

The result object is a list containing multiple elements. The first element is a dataframe of statistics. 
```{r}
head(Res$statistics)
```

We can further determine the dynamic genes (DG) as the genes with fdr.overall $< 0.05$.
```{r}
diffgene <- rownames(Res$statistics)[Res$statistics[,1] < 0.05]
str(diffgene)
```


### population fit
We can estimate the population-level fitting for all DG, which is useful for the downstream analysis.
```{r}
Res$populationFit <- getPopulationFit(Res, gene = diffgene, type = 'time')
```

### clustering
```{r}
Res$cluster <- clusterGene(Res, gene = diffgene, type = 'time', k=3)
```

### plotClusterMean
```{r fig_sct_plotclustermean, fig.height = 2.3, fig.width = 3.3, fig.align = "center"}
plotClusterMean(testobj=Res, cluster = Res$cluster, type = 'time')
```

### GO analysis
```{r eval = FALSE}
goRes <- GOEnrich(testobj = Res, type = 'time', sep = ':.*')
plotGOEnrich(goRes = goRes)
```

### compare original and fitted expression
```{r fig_sct_plotFitHm, fig.height = 4.5, fig.width = 10, fig.align = "center"}
plotFitHm(Res, subsampleCell  = F, showCluster = T, type = 'time', cellWidthTotal = 200, cellHeightTotal = 200)
```
     

## Cell Proportion Test
```{r}
Res <- cell_prop_test(cellanno = expdata$cellanno, pseudotime = expdata$pseudotime, design = expdata$design, ncores = 4)
```

The dataframe `statistics` in the `Res` object contains the test result: a *p*-value `pval.overall` and a effect size *z*-score `z.overall`.
```{r}
Res$statistics
```

#### Session Info

```{r}
sessionInfo()
```

#### Citation

If the `Lamian` package is useful in your work, please cite the following paper:
- (give paper info here)
